### ISSUE TITLE
Shader support

### ISSUE THREAD

**Issue by vlidholt** [cocos2d/cocos2d-objc#360]
Support for shaders, including adding more than one shader and applying shaders on the root level (without a specific render buffer).

Shader support should also add an effects node (CCEffectsNode) which support a number of preset effects/filters.

**Comment by slembcke** [comment]
Including more than one shader? Like doing multi-pass rendering? Not sure I understand what applying a shader to the root level (the scene node?) would do either.

**Comment by Birkemose** [comment]
I have had a lot of mixed feelings on shaders, but I keep returning to "keep it simple".

By being able to apply more than one shader pr node, I understand it as the ability to plug in a new shader, just like you would set the colour.

// assign a shader by name
mySprite.shader = [[CCShaderCache sharedShaderCache] getShaderForName:@"greyscale"];
No additional code should be needed to use the shader.

I would then keep the shaders very simple, and hide all openGL from the user. In case the user needs additional uniforms, they should be inked to the shader when the shader is created, and automatically updated upon render.

// create a new shader, and link a property from parent to a uniform
myShader = [CCShaderCache sharedShaderCache] addShaderForName:(NSString *)shaderName vertexFile:(NSString *)vertFile fragmentFile:(NSString *)fragFile];
[myShader linkUniform:(NSString *)uniformName toProperty:(NSString *)propertyName];

The shader expects a basic attribute setup for a single texture. It would be a killer feature to be able to assign a second or third texture from a CCSprite, directly to the shader, and thus enabling multi texturing. This is "a must have" feature, as soon as we get past the ping-pong games. The syntax could look something like 

[myShader assignSecondTexture:(CCTexture *)texture];

This would automatically set up attributes and uniforms needed for a full second texture implementation. Likewise, I see no reason why not also adding third and fourth texture.

**Comment by Birkemose** [comment]
Btw, I forgot to mention, that I would remove the automatically inserted stuff into the shaders, and in stead rely on some good templates.

**Comment by vlidholt** [comment]
@slembcke what I meant with the root node is to apply the shaders to the frame buffer directly as discussed here: http://www.cocos2d-iphone.org/forums/topic/50-shaders-of-grey/#post-396766 It seems like it can improve performance greatly for full screen effects.

**Comment by Birkemose** [comment]
Another option was to support GLKit effects for now, until we settle on how to do shaders.

**Comment by FramusRock** [comment]
I've read somewhere that an overhaul of the complete rendering is planned anyway: Wouldn't it be a good idea to do that first and care for the best possible shader support while doing it?

**Comment by slembcke** [comment]
Does GLKit provide fullscreen effects? There is the GLKEffect class, but that is more like a surface material for 3D models to aid in porting GL ES 1.1 code. I'm not aware of any other GLKit features that apply here.

**Comment by Birkemose** [comment]
I think it does. At least it enables you to render sprites, without having to worry about shaders.
Aiming for shaders in the december release, will not be possible, so the benefit of using effect nodes would be, that we could kill of shaders completely for next release, and then have the time to incorporate a proper shader for 3.1

That would reduce confusion, and effect nodes still allow people to apply various effects to the sprites.
What does people think about this?

**Comment by billhollings** [comment]
Okay...this is the other big area (after GL state) where cocos3d has some technology that may (or may not) be useful. Although, again...it might be overkill. But let's explore...

The mapping of uniforms to properties is a good idea:

```
[myShader linkUniform:(NSString *)uniformName toProperty:(NSString *)propertyName]
```

In cococs3d, this is done more or less automatically by declaring semantics. But, the needs might be different, and cocos3d's design for that part is probably overkill.

At some point, you'll want to allow someone to configure the uniform-property mapping via a PList or other configuration-based mechanism, so they can write a shader and provide the mapping for their variables to properties without having to write a bunch of lines of code like the above.

You can also introduce default uniforms for each node property. So if a user declares the uniform with a known standard name (eg. u_CCNodeOpacity), the mapping is automatic, and they don't have to write the above lines.

Adding multi-texturing is also a good idea. But it should be done by adding it to the sprite, not the shader. Keep in mind that shaders are reused by many nodes, so your line:

```
[myShader assignSecondTexture:(CCTexture *)texture];
```

effectively hard-codes all nodes that use that shader to use the same texture as the second texture.

Instead, you might want to simply create a CCMutliTextureSprite that can support an addTexture: method and keeps the textures in an array. Or you could declare specific uses, such as a lightMapTexture property, or a bumpMapTexture property on such a multi-texturing node. By doing that, you can possibly automatically match a node to the correct shader. If a node has a bumpMapTexture populated, then it automatically uses the BumpMap.fsh fragment shader, etc.

My advice is also to add _both_ CCShader and CCShaderProgram classes, that are each sparately cached, particularly if you are going to allow loading and compiling them from files. You'll find that the same vertex shader is used over and over in a shader program with different fragment shaders, so it makes sense to load and compile it only once. Load it once, cache it, and link it to different fragment shaders as needed by the different shader programs.

A CCGLSLVariable class, with two subclasses, CCGLSLAttribute and CCGLSLUniform as first-class objects is very useful too, to take on some of the management of your uniform-property mappings described above. The CCGLSLUniform should also take care of managing state (this again!  :^) ), so that when several nodes use the same value for a uniform, it is not set in the shader program over and over.

There are certainly some parts of this that can come from cocos3d, namely the classes: CCShaderProgram, CCShader, CCGLSLUniform and CCGLAttribute, and the whole loading, caching, compiling, and linking mechanism. I think I wouldn't bring over the semantic mappings, and we could just have a CCGLSLUniform carry an invocation object to be invoked on the CCNode to retrieve the property value when it needs to be populated. So, a simpler mechanism than currently is used by cocos3d.

Let me know when you want to wrestle with this, and we can discuss the available cocos3d components in more detail.

**Comment by vlidholt** [comment]
From scott:
Anybody want to start a wish list of shaders?

A lot of GPUImage is highly templated shaders (ex: https://github.com/BradLarson/GPUImage/blob/master/framework/Source/GPUImageGaussianBlurFilter.m) Probably not a lot we can use directly, but it might be worthwhile to see if we can integrate it in as a post-process effects library. Could be really neat.

**Comment by vlidholt** [comment]
So, I'm posting my wish list for what I'd like to see in terms of shaders.
1. A high level of abstraction for those of us that do not want to figure out how to write our own shaders, combined with a rather large library of pre-made shaders. Compare SpriteKit's effect node. I'm thinking things like: Color correction (saturation, contrast, brightness, levels, exposure, hue, monochrome, sepia tone), effects (blur, sharpen, unsharp mask, pixellate, polka dots, half tone), transformations (3d transforms – e.g. for flipping cards, transitions etc).
2. Multi-pass, so that it's possible to apply both blur and desaturate.
3. Possibility to write advanced custom shaders, for those that want to fiddle around with them.
4. Being able to support things like Sprite Lamp: http://www.kickstarter.com/projects/finnmorgan/sprite-lamp-dynamic-lighting-for-2d-art . Ideally, have this as one of the higher level options would be really cool and we could potentially even add SpriteBuilder support for it!

**Comment by billhollings** [comment]
How would you see multi-pass shaders working?

For what it's worth...rather than multi-pass, I've been exploring dynamically built shaders, that could be built by selecting one or more effects, that would then be assembled into a compiled shader.

**Comment by vlidholt** [comment]
So, I'm not sure if this is possible to accomplish. I'm trying to describe what I'd want from a user perspective. Dynamically built shaders could definitely be a solution to this!

It's also possible that this should be split up into two features; one being the shaders and one being some sort of effects node built on top of the shader feature.

My main point here, is for this to be a killer feature, I think that there needs to be a really easy to use API that goes with it. If you could very easily do post-rendering effects and the like, this is something that would make it into very many Cocos2D games.

**Comment by cocojoe** [comment]
I like the idea of an easy API for _tinkering_ that would allow users to
play with shader effects, I envision something along the lines of Particle
Editor.  Be quite nice to have this as a stand alone app and/or plugin tool
for SpriteBuilder.

On 16 January 2014 22:41, Viktor Lidholt notifications@github.com wrote:

> So, I'm not sure if this is possible to accomplish. I'm trying to describe
> what I'd want from a user perspective. Dynamically built shaders could
> definitely be a solution to this!
> 
> It's also possible that this should be split up into two features; one
> being the shaders and one being some sort of effects node built on top of
> the shader feature.
> 
> My main point here, is for this to be a killer feature, I think that there
> needs to be a really easy to use API that goes with it. If you could very
> easily do post-rendering effects and the like, this is something that would
> make it into very many Cocos2D games.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/cocos2d/cocos2d-iphone/issues/360#issuecomment-32555092
> .

## 

Martin

**Comment by cocojoe** [comment]
This was posted on forum, it's quite fun from the user perspective and
something similar for Cocos2D would be quite nice.  I like the idea that
SpriteBuilder could just have an effect button which opens up a panel that
does something similar.

http://dl.dropboxusercontent.com/u/32864004/dev/FPDemo/PunkFX%20latest%20demo.swf

On 17 January 2014 09:39, Martin Walsh martin.walsh@gmail.com wrote:

> I like the idea of an easy API for _tinkering_ that would allow users to
> play with shader effects, I envision something along the lines of Particle
> Editor.  Be quite nice to have this as a stand alone app and/or plugin tool
> for SpriteBuilder.
> 
> On 16 January 2014 22:41, Viktor Lidholt notifications@github.com wrote:
> 
> > So, I'm not sure if this is possible to accomplish. I'm trying to
> > describe what I'd want from a user perspective. Dynamically built shaders
> > could definitely be a solution to this!
> > 
> > It's also possible that this should be split up into two features; one
> > being the shaders and one being some sort of effects node built on top of
> > the shader feature.
> > 
> > My main point here, is for this to be a killer feature, I think that
> > there needs to be a really easy to use API that goes with it. If you could
> > very easily do post-rendering effects and the like, this is something that
> > would make it into very many Cocos2D games.
> > 
> > —
> > Reply to this email directly or view it on GitHubhttps://github.com/cocos2d/cocos2d-iphone/issues/360#issuecomment-32555092
> > .
> 
> ## 
> 
> Martin

## 

Martin

**Comment by Birkemose** [comment]
If we create an array of shaders, and some uniforms to control them, I guess it would be relatively easy to add effects to SpriteBuilder.
Branching in shaders based on uniforms, has very little overhead, so it should be possible to create a highly customisable shader, if ultra high performance is not required.

**Comment by slembcke** [comment]
Uber shaders like that would be useful in a lot of cases. Fullscreen post-process effects have to be relatively fast, but shaders applied to a few small sprites can go really crazy on their complexity as long as their pixel coverage isn't that much.

**Comment by slembcke** [comment]
Most of this has been implemented in the v3.1 branch now or has been spun off into the CCEffect idea.
